File: tests/support/outcomes.py
from __future__ import annotations

from typing import Any
import unittest

from incident_triage_agent.domain import (
    IncidentClass,
    NextAction,
    SafetyStatus,
    SourceTier,
    TriageRun,
    WorkflowState,
)


def assert_valid_run_outcome(
    case: unittest.TestCase,
    run: TriageRun,
    *,
    incident_class: IncidentClass | str | None = None,
    next_action: NextAction | str | None = None,
    evidence_prefixes: tuple[str, ...] = (),
    cited_sources: tuple[str, ...] = (),
    cited_tiers: tuple[SourceTier | str, ...] = (),
    safety_status: SafetyStatus | str | None = None,
    approval_required: bool | None = None,
    scorecard_checks: tuple[str, ...] = (),
) -> None:
    case.assertIsNotNone(run.validation, "expected run to include validation result")
    assert run.validation is not None
    case.assertTrue(run.validation.valid, f"expected valid decision, got errors={run.validation.errors!r}")
    case.assertIsNotNone(run.validation.decision, "expected valid run to include a decision")
    assert run.validation.decision is not None
    case.assertIn(WorkflowState.DECISION_VALIDATED, run.states, "expected decision_validated state")
    case.assertIn(WorkflowState.SCORED, run.states, "expected scored state")

    decision = run.validation.decision
    if incident_class is not None:
        case.assertEqual(
            _value(incident_class),
            decision.incident_class.value,
            f"expected incident_class={_value(incident_class)!r}, got {decision.incident_class.value!r}",
        )
    if next_action is not None:
        case.assertEqual(
            _value(next_action),
            decision.next_action.value,
            f"expected next_action={_value(next_action)!r}, got {decision.next_action.value!r}",
        )

    _assert_cited_prefixes(case, tuple(decision.evidence_ids), evidence_prefixes)

    case.assertIsNotNone(run.evidence_package, "expected run to include evidence package")
    assert run.evidence_package is not None
    provenance = run.evidence_package.provenance_summary(decision.evidence_ids)
    _assert_contains_all(
        case,
        tuple(source.value if hasattr(source, "value") else source for source in provenance.cited_tiers),
        tuple(_value(tier) for tier in cited_tiers),
        "cited_tiers",
    )
    _assert_contains_all(case, provenance.cited_sources, cited_sources, "cited_sources")

    if safety_status is not None or approval_required is not None:
        case.assertIsNotNone(run.safety, "expected run to include safety result")
        assert run.safety is not None
        _assert_safety(case, run.safety.status, run.safety.approval_required, safety_status, approval_required)
        if run.safety.approval_required:
            case.assertIsNotNone(run.safety.staged_payload, "expected approval-required outcome to include staged payload")
            case.assertIsNotNone(run.safety.audit_event, "expected approval-required outcome to include audit event")
            assert run.safety.audit_event is not None
            case.assertFalse(run.safety.audit_event.get("executed"), "expected staged action to remain unexecuted")

    _assert_scorecard_checks(case, run.scorecard.scores if run.scorecard else None, scorecard_checks)


def assert_valid_response_outcome(
    case: unittest.TestCase,
    response: dict[str, Any],
    *,
    incident_class: IncidentClass | str | None = None,
    next_action: NextAction | str | None = None,
    evidence_prefixes: tuple[str, ...] = (),
    cited_sources: tuple[str, ...] = (),
    available_tiers: tuple[SourceTier | str, ...] = (),
    cited_tiers: tuple[SourceTier | str, ...] = (),
    require_safety: bool = False,
    safety_status: SafetyStatus | str | None = None,
    approval_required: bool | None = None,
    scorecard_checks: tuple[str, ...] = (),
) -> None:
    case.assertEqual(response.get("status"), "ok", f"expected ok response, got {response!r}")
    validation = response.get("validation")
    case.assertIsInstance(validation, dict, "expected response validation object")
    assert isinstance(validation, dict)
    case.assertTrue(validation.get("valid"), f"expected valid response, got errors={validation.get('errors')!r}")
    decision = response.get("decision")
    case.assertIsInstance(decision, dict, "expected valid response to include decision")
    assert isinstance(decision, dict)

    if incident_class is not None:
        case.assertEqual(
            _value(incident_class),
            decision.get("incident_class"),
            f"expected incident_class={_value(incident_class)!r}, got {decision.get('incident_class')!r}",
        )
    if next_action is not None:
        case.assertEqual(
            _value(next_action),
            decision.get("next_action"),
            f"expected next_action={_value(next_action)!r}, got {decision.get('next_action')!r}",
        )

    _assert_cited_prefixes(case, tuple(decision.get("evidence_ids", ())), evidence_prefixes)

    provenance = response.get("provenance")
    case.assertIsInstance(provenance, dict, "expected response provenance object")
    assert isinstance(provenance, dict)
    _assert_contains_all(case, tuple(provenance.get("available_tiers", ())), tuple(_value(tier) for tier in available_tiers), "available_tiers")
    _assert_contains_all(case, tuple(provenance.get("cited_tiers", ())), tuple(_value(tier) for tier in cited_tiers), "cited_tiers")
    _assert_contains_all(case, tuple(provenance.get("cited_sources", ())), cited_sources, "cited_sources")

    if require_safety or safety_status is not None or approval_required is not None:
        safety = response.get("safety")
        case.assertIsInstance(safety, dict, "expected response safety object")
        assert isinstance(safety, dict)
        _assert_safety(case, safety.get("status"), safety.get("approval_required"), safety_status, approval_required)
        audit_event = safety.get("audit_event")
        if audit_event is not None:
            case.assertNotEqual(audit_event.get("executed"), True, "expected safety result not to execute actions")
        if safety.get("approval_required"):
            case.assertIsInstance(audit_event, dict, "expected approval-required response to include audit event")

    scorecard = response.get("scorecard")
    _assert_scorecard_checks(
        case,
        scorecard.get("scores") if isinstance(scorecard, dict) else None,
        scorecard_checks,
    )


def assert_recoverable_run(
    case: unittest.TestCase,
    run: TriageRun,
    *,
    error_contains: str | None = None,
) -> None:
    case.assertIn(WorkflowState.RECOVERABLE_FAILURE, run.states, "expected recoverable_failure state")
    case.assertIn(WorkflowState.SCORED, run.states, "expected scored state")
    case.assertIsNotNone(run.validation, "expected validation errors on recoverable run")
    assert run.validation is not None
    case.assertFalse(run.validation.valid, "expected recoverable run validation to be invalid")
    case.assertIsNone(run.validation.decision, "expected recoverable run to omit trusted decision")
    if error_contains is not None:
        case.assertTrue(
            any(error_contains in error for error in run.validation.errors),
            f"expected validation error containing {error_contains!r}, got {run.validation.errors!r}",
        )
    case.assertIsNone(run.safety, "expected recoverable run to skip safety action")
    case.assertIsNotNone(run.scorecard, "expected recoverable run to include scorecard")


def assert_recoverable_response(
    case: unittest.TestCase,
    response: dict[str, Any],
    *,
    error_contains: str | None = None,
) -> None:
    case.assertEqual(response.get("status"), "ok", f"expected ok response, got {response!r}")
    case.assertIn(WorkflowState.RECOVERABLE_FAILURE.value, response.get("states", ()), "expected recoverable_failure state")
    validation = response.get("validation")
    case.assertIsInstance(validation, dict, "expected response validation object")
    assert isinstance(validation, dict)
    case.assertFalse(validation.get("valid"), "expected response validation to be invalid")
    case.assertNotIn("decision", response, "expected recoverable response to omit trusted decision")
    case.assertNotIn("safety", response, "expected recoverable response to skip safety action")
    if error_contains is not None:
        errors = tuple(validation.get("errors", ()))
        case.assertTrue(
            any(error_contains in error for error in errors),
            f"expected validation error containing {error_contains!r}, got {errors!r}",
        )
    case.assertIn("scorecard", response, "expected recoverable response to include scorecard")


def assert_ignored_response(
    case: unittest.TestCase,
    status: int,
    response: dict[str, Any],
    *,
    reason: str,
) -> None:
    case.assertEqual(status, 202)
    case.assertEqual(response.get("status"), "ignored")
    case.assertEqual(response.get("reason"), reason)
    case.assertNotIn("decision", response)
    case.assertNotIn("safety", response)


def _assert_cited_prefixes(
    case: unittest.TestCase,
    evidence_ids: tuple[str, ...],
    evidence_prefixes: tuple[str, ...],
) -> None:
    for prefix in evidence_prefixes:
        case.assertTrue(
            any(evidence_id.startswith(prefix) for evidence_id in evidence_ids),
            f"expected cited evidence prefix {prefix!r}, got {evidence_ids!r}",
        )


def _assert_contains_all(
    case: unittest.TestCase,
    actual: tuple[str, ...],
    expected: tuple[str, ...],
    label: str,
) -> None:
    for item in expected:
        case.assertIn(item, actual, f"expected {label} to include {item!r}, got {actual!r}")


def _assert_safety(
    case: unittest.TestCase,
    actual_status: Any,
    actual_approval_required: Any,
    expected_status: SafetyStatus | str | None,
    expected_approval_required: bool | None,
) -> None:
    if expected_status is not None:
        case.assertEqual(
            _value(expected_status),
            actual_status,
            f"expected safety_status={_value(expected_status)!r}, got {actual_status!r}",
        )
    if expected_approval_required is not None:
        case.assertEqual(
            expected_approval_required,
            actual_approval_required,
            f"expected approval_required={expected_approval_required!r}, got {actual_approval_required!r}",
        )


def _assert_scorecard_checks(
    case: unittest.TestCase,
    scores: dict[str, bool] | None,
    scorecard_checks: tuple[str, ...],
) -> None:
    if not scorecard_checks:
        return
    case.assertIsNotNone(scores, "expected scorecard scores")
    assert scores is not None
    for check in scorecard_checks:
        case.assertIn(check, scores, f"expected scorecard check {check!r}, got {sorted(scores)!r}")
        case.assertTrue(scores[check], f"expected scorecard check {check!r} to pass")


def _value(value: Any) -> str:
    if hasattr(value, "value"):
        return str(value.value)
    return str(value)

File: tests/support/outcomes.ts
import { expect } from "vitest";
import type { IncidentClass, NextAction, SourceTier, WorkflowState } from "../../src/domain";
import type { TriageRun } from "../../src/workflow";
import type { SafetyStatus } from "../../src/policy";

export function assertValidRunOutcome(
  run: TriageRun,
  options: {
    incidentClass?: IncidentClass | string;
    nextAction?: NextAction | string;
    evidencePrefixes?: string[];
    citedSources?: string[];
    citedTiers?: Array<SourceTier | string>;
    safetyStatus?: SafetyStatus | string;
    approvalRequired?: boolean;
    scorecardChecks?: string[];
  } = {},
): void {
  expect(run.validation, "expected run to include validation result").toBeDefined();
  expect(run.validation?.valid, `expected valid decision, got errors=${JSON.stringify(run.validation?.errors)}`).toBe(true);
  expect(run.validation?.decision, "expected valid run to include a decision").toBeDefined();
  expect(run.states, "expected decision_validated state").toContain("decision_validated" satisfies WorkflowState);
  expect(run.states, "expected scored state").toContain("scored" satisfies WorkflowState);

  const decision = run.validation?.decision;
  if (!decision) {
    throw new Error("expected decision");
  }
  if (options.incidentClass !== undefined) {
    expect(decision.incidentClass).toBe(value(options.incidentClass));
  }
  if (options.nextAction !== undefined) {
    expect(decision.nextAction).toBe(value(options.nextAction));
  }

  assertCitedPrefixes(decision.evidenceIds, options.evidencePrefixes ?? []);

  expect(run.evidencePackage, "expected run to include evidence package").toBeDefined();
  const provenance = run.evidencePackage?.provenanceSummary(decision.evidenceIds);
  expect(provenance).toBeDefined();
  if (!provenance) {
    throw new Error("expected provenance");
  }
  assertContainsAll(provenance.citedTiers.map(value), (options.citedTiers ?? []).map(value), "cited_tiers");
  assertContainsAll(provenance.citedSources, options.citedSources ?? [], "cited_sources");

  if (options.safetyStatus !== undefined || options.approvalRequired !== undefined) {
    expect(run.safety, "expected run to include safety result").toBeDefined();
    assertSafety(run.safety?.status, run.safety?.approvalRequired, options.safetyStatus, options.approvalRequired);
    if (run.safety?.approvalRequired) {
      expect(run.safety.stagedPayload, "expected approval-required outcome to include staged payload").toBeDefined();
      expect(run.safety.auditEvent, "expected approval-required outcome to include audit event").toBeDefined();
      expect(run.safety.auditEvent?.executed, "expected staged action to remain unexecuted").toBe(false);
    }
  }

  assertScorecardChecks(run.scorecard?.scores, options.scorecardChecks ?? []);
}

export function assertValidResponseOutcome(
  response: Record<string, any>,
  options: {
    incidentClass?: IncidentClass | string;
    nextAction?: NextAction | string;
    evidencePrefixes?: string[];
    citedSources?: string[];
    availableTiers?: Array<SourceTier | string>;
    citedTiers?: Array<SourceTier | string>;
    requireSafety?: boolean;
    safetyStatus?: SafetyStatus | string;
    approvalRequired?: boolean;
    scorecardChecks?: string[];
  } = {},
): void {
  expect(response.status, `expected ok response, got ${JSON.stringify(response)}`).toBe("ok");
  expect(response.validation, "expected response validation object").toEqual(expect.any(Object));
  expect(response.validation.valid, `expected valid response, got errors=${JSON.stringify(response.validation.errors)}`).toBe(true);
  expect(response.decision, "expected valid response to include decision").toEqual(expect.any(Object));

  if (options.incidentClass !== undefined) {
    expect(response.decision.incident_class).toBe(value(options.incidentClass));
  }
  if (options.nextAction !== undefined) {
    expect(response.decision.next_action).toBe(value(options.nextAction));
  }

  assertCitedPrefixes(response.decision.evidence_ids ?? [], options.evidencePrefixes ?? []);

  expect(response.provenance, "expected response provenance object").toEqual(expect.any(Object));
  assertContainsAll(response.provenance.available_tiers ?? [], (options.availableTiers ?? []).map(value), "available_tiers");
  assertContainsAll(response.provenance.cited_tiers ?? [], (options.citedTiers ?? []).map(value), "cited_tiers");
  assertContainsAll(response.provenance.cited_sources ?? [], options.citedSources ?? [], "cited_sources");

  if (options.requireSafety || options.safetyStatus !== undefined || options.approvalRequired !== undefined) {
    expect(response.safety, "expected response safety object").toEqual(expect.any(Object));
    assertSafety(response.safety.status, response.safety.approval_required, options.safetyStatus, options.approvalRequired);
    if (response.safety.audit_event !== undefined) {
      expect(response.safety.audit_event.executed, "expected safety result not to execute actions").not.toBe(true);
    }
    if (response.safety.approval_required) {
      expect(response.safety.audit_event, "expected approval-required response to include audit event").toEqual(expect.any(Object));
    }
  }

  assertScorecardChecks(response.scorecard?.scores, options.scorecardChecks ?? []);
}

export function assertRecoverableRun(run: TriageRun, errorContains?: string): void {
  expect(run.states, "expected recoverable_failure state").toContain("recoverable_failure" satisfies WorkflowState);
  expect(run.states, "expected scored state").toContain("scored" satisfies WorkflowState);
  expect(run.validation, "expected validation errors on recoverable run").toBeDefined();
  expect(run.validation?.valid, "expected recoverable run validation to be invalid").toBe(false);
  expect(run.validation?.decision, "expected recoverable run to omit trusted decision").toBeUndefined();
  if (errorContains !== undefined) {
    expect(run.validation?.errors.some((error) => error.includes(errorContains))).toBe(true);
  }
  expect(run.safety, "expected recoverable run to skip safety action").toBeUndefined();
  expect(run.scorecard, "expected recoverable run to include scorecard").toBeDefined();
}

export function assertRecoverableResponse(response: Record<string, any>, errorContains?: string): void {
  expect(response.status, `expected ok response, got ${JSON.stringify(response)}`).toBe("ok");
  expect(response.states, "expected recoverable_failure state").toContain("recoverable_failure");
  expect(response.validation, "expected response validation object").toEqual(expect.any(Object));
  expect(response.validation.valid, "expected response validation to be invalid").toBe(false);
  expect(response, "expected recoverable response to omit trusted decision").not.toHaveProperty("decision");
  expect(response, "expected recoverable response to skip safety action").not.toHaveProperty("safety");
  if (errorContains !== undefined) {
    expect((response.validation.errors ?? []).some((error: string) => error.includes(errorContains))).toBe(true);
  }
  expect(response, "expected recoverable response to include scorecard").toHaveProperty("scorecard");
}

export function assertIgnoredResponse(status: number, response: Record<string, unknown>, reason: string): void {
  expect(status).toBe(202);
  expect(response.status).toBe("ignored");
  expect(response.reason).toBe(reason);
  expect(response).not.toHaveProperty("decision");
  expect(response).not.toHaveProperty("safety");
}

function assertCitedPrefixes(evidenceIds: string[], prefixes: string[]): void {
  for (const prefix of prefixes) {
    expect(
      evidenceIds.some((evidenceId) => evidenceId.startsWith(prefix)),
      `expected cited evidence prefix ${prefix}, got ${evidenceIds.join(", ")}`,
    ).toBe(true);
  }
}

function assertContainsAll(actual: string[], expected: string[], label: string): void {
  for (const item of expected) {
    expect(actual, `expected ${label} to include ${item}, got ${actual.join(", ")}`).toContain(item);
  }
}

function assertSafety(
  actualStatus: unknown,
  actualApprovalRequired: unknown,
  expectedStatus?: SafetyStatus | string,
  expectedApprovalRequired?: boolean,
): void {
  if (expectedStatus !== undefined) {
    expect(actualStatus).toBe(value(expectedStatus));
  }
  if (expectedApprovalRequired !== undefined) {
    expect(actualApprovalRequired).toBe(expectedApprovalRequired);
  }
}

function assertScorecardChecks(scores: Record<string, boolean> | undefined, checks: string[]): void {
  if (checks.length === 0) {
    return;
  }
  expect(scores, "expected scorecard scores").toBeDefined();
  for (const check of checks) {
    expect(scores, `expected scorecard check ${check}`).toHaveProperty(check);
    expect(scores?.[check], `expected scorecard check ${check} to pass`).toBe(true);
  }
}

function value(input: { value: string } | string): string {
  return typeof input === "string" ? input : input.value;
}
